libnmstate/nm/dns.py

```diff
@@ -0,0 +1,104 @@
+from operator import methodcaller
+
+from libnmstate import iplib
+from libnmstate.nm import nmclient
+from libnmstate.schema import DNS
+from libnmstate.nm import ipv4 as nm_ipv4
+from libnmstate.nm import ipv6 as nm_ipv6
+from libnmstate.error import NmstateInternalError
+
+
+DNS_DEFAULT_PRIORITY_VPN = 50
+DNS_DEFAULT_PRIORITY_OTHER = 100
+
+IPV6_ADDRESS_LENGTH = 128
+
+
+def get_running():
+    dns_state = {
+        DNS.SERVER: [],
+        DNS.SEARCH: []
+    }
+    client = nmclient.client()
+    dns_confs = client.get_dns_configuration()
+    for dns_conf in sorted(dns_confs, key=methodcaller('get_priority')):
+        iface_name = dns_conf.get_interface()
+        for ns in dns_conf.get_nameservers():
+            if iplib.is_ipv6_link_local_addr(ns, IPV6_ADDRESS_LENGTH):
+                if not iface_name:
+                    # For IPv6 link local address, the interface name should be
+                    # appended also.
+                    raise NmstateInternalError(
+                        "Got empty interface name for IPv6 link local "
+                        "DNS nameserver entry {}".format(ns))
+                ns_addr = "{}%{}".format(ns, iface_name)
+            else:
+                ns_addr = ns
+            dns_state[DNS.SERVER].append(ns_addr)
+        dns_state[DNS.SEARCH].extend(dns_conf.get_domains())
+    if not dns_state[DNS.SERVER] and dns_state[DNS.SEARCH]:
+        return {}
+    return dns_state
+
+
+def get_config():
+    dns_conf = {
+        DNS.SERVER: [],
+        DNS.SEARCH: []
+    }
+    tmp_dns_confs = []
+    client = nmclient.client()
+    for ac in client.get_active_connections():
+        # NM prefers IPv6 over IPv4 DNS.
+        for ip_profile in (nm_ipv6.get_ip_profile(ac),
+                           nm_ipv4.get_ip_profile(ac)):
+            if not ip_profile:
+                continue
+            if not ip_profile.props.dns and not ip_profile.props.dns_search:
+                continue
+            priority = ip_profile.props.dns_priority
+            if priority == 0:
+                if ac.get_vpn():
+                    priority = DNS_DEFAULT_PRIORITY_VPN
+                else:
+                    priority = DNS_DEFAULT_PRIORITY_OTHER
+
+            tmp_dns_confs.append(
+                {
+                    'server': ip_profile.props.dns,
+                    'priority': priority,
+                    'search': ip_profile.props.dns_search,
+                }
+            )
+    # Even NM also sort DNS entries base on whether action connection is
+    # holding default gateway, but that is runtime status which could generate
+    # different results. Hence, we just sort in priority and when apply back we
+    # will use different priority to enforce this ordering.
+    tmp_dns_confs.sort(key=itemgetter('priority'))
+    dns_conf = {
+        DNS.SERVER: [],
+        DNS.SEARCH: []
+    }
+    for e in tmp_dns_confs:
+        dns_conf[DNS.SERVER].extend(e['server'])
```

Comment:
`server` and `search` should be constants.

Reply:
This is internal dict, not the public facing data. The constants is for DNS state, not internal dict.